### PR TITLE
Feature/19873 ruby in docker

### PR DESCRIPTION
**Pull request by GuyHarwood**
- Full test suite runs in containers, via docker compose
- Migrations run after SQL ready and before admin app starts
- Test container exit codes influence build exit code
- Cucumber reports & screenshots available as artifacts for download
- Docker Layer caching optimised for quicker build times

### Note: Ignore feature files and ruby tests in this PR under `./test/pupil` and `./test/admin` 
these are just a copy of the ones under the app folders.  the ones under the app folders will be removed in a separate piece of work by @activemq & @sinhaamit1980

### CODE REVIEW

File: test/admin/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+# throw errors if Gemfile has been modified since Gemfile.lock
+RUN bundle config --global frozen 1
+
+# Update the repositories
+# Install utilities
+# Install XVFB and TinyWM
+# Install fonts
+# Install Python
+RUN apt-get -yqq update && \
+    apt-get -yqq install curl unzip && \
```
Comment (jon-shipley) [inline test/admin/Dockerfile]: Is there a reason not to add all the `install` packages over a single `apt-get` command?

File: test/admin/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    curl -sS -o /tmp/chromedriver_linux64.zip http://chromedriver.storage.googleapis.com/$CHROMEDRIVER_VERSION/chromedriver_linux64.zip && \
+    unzip -qq /tmp/chromedriver_linux64.zip -d /opt/chromedriver-$CHROMEDRIVER_VERSION && \
+    rm /tmp/chromedriver_linux64.zip && \
+    chmod +x /opt/chromedriver-$CHROMEDRIVER_VERSION/chromedriver && \
+    ln -fs /opt/chromedriver-$CHROMEDRIVER_VERSION/chromedriver /usr/local/bin/chromedriver
+
+# Install Google Chrome
+RUN curl -sS -o - https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add - && \
+    echo "deb http://dl.google.com/linux/chrome/deb/ stable main" >> /etc/apt/sources.list.d/google-chrome.list && \
+    apt-get -yqq update && \
```
Comment (jon-shipley) [inline test/admin/Dockerfile]: `apt-get update`  was already run on line 14...

File: test/admin/Dockerfile (hunk 1 above)
Comment (GuyHarwood) [reply test/admin/Dockerfile]: it's just easier to read IMO.

File: test/admin/Dockerfile (hunk 2 above)
Comment (GuyHarwood) [reply test/admin/Dockerfile]: removed, good spot!

**Pull request merged**